enderio-machines/src/main/java/com/enderio/machines/common/blockentity/AlloySmelterBlockEntity.java

```diff
@@ -0,0 +1,189 @@
+import com.enderio.machines.common.MachineTier;
+import com.enderio.machines.common.blockentity.base.PoweredCraftingMachineEntity;
+import com.enderio.machines.common.blockentity.data.sidecontrol.item.ItemHandlerMaster;
+import com.enderio.machines.common.blockentity.data.sidecontrol.item.ItemSlotLayout;
+import com.enderio.machines.common.machine.AlloySmelterMode;
+import com.enderio.machines.common.menu.AlloySmelterMenu;
+import com.enderio.machines.common.recipe.AlloySmeltingRecipe;
+import com.enderio.machines.common.recipe.MachineRecipes;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.NonNullList;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.world.Container;
+import net.minecraft.world.entity.player.Inventory;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.inventory.AbstractContainerMenu;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.item.crafting.Recipe;
+import net.minecraft.world.item.crafting.RecipeType;
+import net.minecraft.world.item.crafting.SmeltingRecipe;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.minecraft.world.level.block.state.BlockState;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Optional;
+
+// Using the base Recipe<Container> class so we can support both smelting recipe and alloy smelting recipe.
+public class AlloySmelterBlockEntity extends PoweredCraftingMachineEntity<Recipe<Container>> {
+
+    public static class Simple extends AlloySmelterBlockEntity {
+        public Simple(BlockEntityType<?> pType, BlockPos pWorldPosition, BlockState pBlockState) {
+            super(AlloySmelterMode.Alloys, MachineTier.Simple, pType, pWorldPosition, pBlockState);
+        }
+    }
+
+    public static class Furnace extends AlloySmelterBlockEntity {
+        public Furnace(BlockEntityType<?> pType, BlockPos pWorldPosition, BlockState pBlockState) {
+            super(AlloySmelterMode.Furnace, MachineTier.Simple, pType, pWorldPosition, pBlockState);
+        }
+    }
+
+    public static class Standard extends AlloySmelterBlockEntity {
+        public Standard(BlockEntityType<?> pType, BlockPos pWorldPosition, BlockState pBlockState) {
+            super(AlloySmelterMode.All, MachineTier.Standard, pType, pWorldPosition, pBlockState);
+        }
+    }
+
+    private final AlloySmelterMode defaultMode;
+    private AlloySmelterMode mode;
+
+    public AlloySmelterBlockEntity(AlloySmelterMode defaultMode, MachineTier tier, BlockEntityType<?> pType, BlockPos pWorldPosition, BlockState pBlockState) {
+        super(tier, pType, pWorldPosition, pBlockState);
+        this.defaultMode = defaultMode;
+        this.mode = defaultMode;
+
+        // This can be changed by the gui for the normal and enhanced machines.
+        add2WayDataSlot(new EnumDataSlot<>(this::getMode, this::setMode, SyncMode.GUI));
```

Comment:
add check for tier, before adding dataslot so this can never happen, even if the client uses a custom build jar.

Reply:
or add a check in the setMode method

Reply:
The getMode() method ignores the custom mode if the tier is too low to be fair

Reply:
Yeah, but I would still add a check for it